### ISSUE TITLE
broken backwards compatibility

### ISSUE THREAD

**Issue by XSven** [hkoba/p5-File-AddInc#3]
I have upgraded File::AddInc from version 0.001 to 0.002. At our end we have several programs with a BEGIN block like this

`BEGIN {
  use File::AddInc;
  unshift @INC, $INC[0] . '/perl5';
}`

After the upgrade this block does not work anymore because libdir was appended to `@INC` instead of prepended. In the following example we wanted to search for the perl5 directory inside the directory /home/edi/mlib/KNIB-Lib-EdiStarClients. It works with the 0.001 version but it does not work with the 0.002 version anymore.

```
!!!!!!File::AddInc 0.002 @INC:
/ib/proc/kn/source/perl
lib
xlib
/home/micsw/perl5/lib/perl5/5.14.2/aix-thread-multi-64all
/home/micsw/perl5/lib/perl5/5.14.2
/home/micsw/perl5/lib/perl5/aix-thread-multi-64all
/home/micsw/perl5/lib/perl5
/home/micsw/perl-toolchain/lib/perl5/5.14.2/aix-thread-multi-64all
/home/micsw/perl-toolchain/lib/perl5/5.14.2
/home/micsw/perl-toolchain/lib/perl5/aix-thread-multi-64all
/home/micsw/perl-toolchain/lib/perl5
/home/micsw/perl-local-lib/lib/perl5/5.14.2/aix-thread-multi-64all
/home/micsw/perl-local-lib/lib/perl5/5.14.2
/home/micsw/perl-local-lib/lib/perl5/aix-thread-multi-64all
/home/micsw/perl-local-lib/lib/perl5
/usr/local/perl-5.14.2_64_8/lib/site_perl/5.14.2/aix-thread-multi-64all
/usr/local/perl-5.14.2_64_8/lib/site_perl/5.14.2
/usr/local/perl-5.14.2_64_8/lib/5.14.2/aix-thread-multi-64all
/usr/local/perl-5.14.2_64_8/lib/5.14.2
.
/home/edi/mlib/KNIB-Lib-EdiStarClients


!!!!!!File::AddInc 0.001 @INC:
/home/edi/mlib/KNIB-Lib-EdiStarClients
/ib/proc/kn/source/perl
lib
xlib
/home/micsw/perl-toolchain/lib/perl5/5.14.2/aix-thread-multi-64all
/home/micsw/perl-toolchain/lib/perl5/5.14.2
/home/micsw/perl-toolchain/lib/perl5/aix-thread-multi-64all
/home/micsw/perl-toolchain/lib/perl5
/home/micsw/perl-local-lib/lib/perl5/5.14.2/aix-thread-multi-64all
/home/micsw/perl-local-lib/lib/perl5/5.14.2
/home/micsw/perl-local-lib/lib/perl5/aix-thread-multi-64all
/home/micsw/perl-local-lib/lib/perl5
/usr/local/perl-5.14.2_64_8/lib/site_perl/5.14.2/aix-thread-multi-64all
/usr/local/perl-5.14.2_64_8/lib/site_perl/5.14.2
/usr/local/perl-5.14.2_64_8/lib/5.14.2/aix-thread-multi-64all
/usr/local/perl-5.14.2_64_8/lib/5.14.2
.
```

**Comment by hkoba** [comment]
Thank you for telling me about it. I guess your script already added `/home/edi/mlib/KNIB-Lib-EdiStarClients` before the `use File::AddInc` line.
To clarify what is happening there, please try to run your script with setting environment variable `DEBUG_MOP4IMPORT` to 1. You will get logging messages from File::AddInc in STDERR. It will tell you how libdir is treated (added to `@INC` or just ignored).

With the commit b9857d22e3a9240755c419d9a3d072e5e30108ac in 0.002, I changed File::AddInc to avoid *reordering* of existing paths in `@INC` because such reordering can make module loadings less consistent and lead hard to understand bugs.

I'm ok to improve `File::AddInc` to provide previous behavior as another pragma like followings(pragma name suggestions are welcomed).  If I implement it, can it help to solve your problems?
```perl
use File::AddInc -force;
```

**Comment by XSven** [comment]
Amazing you are right! With debugging enabled the 0.001 version says

`# use lib '/home/edi/mlib/KNIB-Lib-EdiStarClients'`

but the 0.002 version says

`# No need to add libs: '/home/edi/mlib/KNIB-Lib-EdiStarClients'`

I will check internally what is going on. For the time being do NOT implement any additional flags like `-force`.

**Comment by XSven** [comment]
Could you please make the message

`# No need to add libs: '...'`

a permanent warning message instead of debug message that is hidden unless `DEBUG_MOP4IMPORT` is enabled. I think trying to add the same directory to `@INC` is most probably a mistake and therefore at least a warning should be raised.

**Comment by hkoba** [comment]
Hi.

I can understand your frustration. Unfortunately, simply changing the debug message to a permanent warning can harm the following use case and not acceptable. 

Please imagine: You have a libdir and have 2 (or more) modulinos `M1.pm` and `M2.pm`.  `M1.pm` uses `M2.pm`. Since both of them are modulino, they use File::AddInc. 

```perl
# ---------------------------
# M1.pm
# ---------------------------
#!/usr/bin/env  perl
package M1;
use File::AddInc;
use M2;
...
1;

# ---------------------------
# M2.pm
# ---------------------------
#!/usr/bin/env  perl
package M2;
use File::AddInc;
...
1;
```

Obviously, in this use case, when you run `M1.pm`, File::AddInc is used twice to add the libdir to `@INC`. This is a completely legal use of File::AddInc. And importantly, I want to encourage such a style of software building (building a set of complex systems as a collection of independently invocable small modules). That is why I wrote File::AddInc.

In other words, File::AddInc is intentionally designed to be [idempotent](https://en.wikipedia.org/wiki/Idempotence#Computer_science_meaning) for the addition of libdir. File::AddInc is designed to ensure libdir exists in `@INC`, not to alter/reorder it repetitively.
